### PR TITLE
CI: add crate publication workflow

### PR DESCRIPTION
**Pull request by zoedberg**
refers to #142 

from the repo's github page go to `Settings -> Secrets` and add a new secret named `CRATES_IO_TOKEN` and, as its value, add an access token from your crates.io account (the one you use for `cargo login <access_token>`)

### CODE REVIEW

**Review by dr-orlovsky — COMMENTED** [review]
Very nice, thank you! But I'm afraid it will not work. When you tag, main crate must reference new unpublished version of derive, which will prevent derive from being published :( (cargo checks man wen publishes derive and fails on that)

**Comment by zoedberg** [comment]
> Very nice, thank you! But I'm afraid it will not work. When you tag, main crate must reference new unpublished version of derive, which will prevent derive from being published :( (cargo checks man wen publishes derive and fails on that)

So normally how do you publish these packages?

Let me guess, but ignore me if I'm wrong (I don't want to add confusion):
in the main `Cargo.toml` you put `lnpbp_derive = { path = "derive" }` under `[dependencies]` and then, from the `derive` directory, you run `cargo publish`. afterwards, in the main `Cargo.toml`, you change the dependency to `lnpbp_derive = "newly_published_version"` and run `cargo publish` (from the project root)

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+env:
+  CARGO_TERM_COLOR: always
+
+jobs:
+  publication:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - run: cargo login $CRATES_IO_TOKEN
```
[the remark this answers is not in the document]
Comment (zoedberg) [reply .github/workflows/publish.yml:16]: I thought it wasn't possible since I had some issues using `${{ env.VAR }}` syntax, but I've just tried and it seems to work, so I'll change it

**Comment by dr-orlovsky** [comment]
I had answered your question, but now I see that the answer is not here. Have you received it?

**Comment by zoedberg** [comment]
I think your answer is [here](https://github.com/LNP-BP/rgb-node/pull/98#issuecomment-724188087)

**Comment by zoedberg** [comment]
If the project structure can't change from the current one, I think we have 2 options:
- automate what you normally do locally, patching the `Cargo.toml` files in the CI workflow
- use a tool like [cargo-workspaces](https://github.com/pksunkara/cargo-workspaces) (probably we'll need to change the way we trigger the CI, if you find this option interesting I can expand how I think we can use it)

**Comment by dr-orlovsky** [comment]
@zoedberg I find workspace option interesting

**Comment by zoedberg** [comment]
@dieeasy would you like to proceed investigating `cargo-workspaces` solution? (I'm a bit overloaded right now)

**Comment by dieeasy** [comment]
#### TL;DR
after several tests with `cargo ws` I can't get it to work properly but it does indeed work while testing on a brand new project

#### test on new project
I'll start by describing the working condition: I created a new project from scratch and used `cargo ws` (`init`, `create`) to setup the basics; following cargo ws's readme and trying to do nothing fancy I ended up with a main `Cargo.toml` that only declares the workspace plus two crates (each in its subdir) where one depends on the other crate; here's the project structure:
```
.
├── Cargo.lock
├── Cargo.toml
├── one
│   ├── Cargo.toml
│   └── src
│       └── main.rs
└── two
    ├── Cargo.toml
    └── src
        └── lib.rs
```

the dependency from one to two has been specified as `two = { path = "../two", version = "0.0.0" }`, following what the author of cargo ws does in the [`clap`](https://github.com/clap-rs/clap/blob/master/Cargo.toml) package (given the readme says "I made this to work on clap")

with this in place, calling `cargo ws version patch` or `cargo ws publish patch` bumps all crates from `0.0.0` to `0.0.1` and also updates the dependency from one to two

command output:
```
info current common version 0.0.0

Changes:
 - one: 0.0.0 => 0.0.1
 - two: 0.0.0 => 0.0.1

✔ Are you sure you want to create these versions? · yes
    Updating one v0.0.0 (/home/dieeasy/test/cargo-ws/one) -> v0.0.1
    Updating two v0.0.0 (/home/dieeasy/test/cargo-ws/two) -> v0.0.1
info success ok
```

resulting diff:
```diff
diff --git a/one/Cargo.toml b/one/Cargo.toml
index acb8461..752557b 100644
--- a/one/Cargo.toml
+++ b/one/Cargo.toml
@@ -3 +3 @@ name = "one"
-version = "0.0.0"
+version = "0.0.1"
@@ -13 +13 @@ name = "one"
-two = { path = "../two", version = "0.0.0" }
+two = { path = "../two", version = "0.0.1" }
diff --git a/two/Cargo.toml b/two/Cargo.toml
index 9c2dd78..af6182b 100644
--- a/two/Cargo.toml
+++ b/two/Cargo.toml
@@ -3 +3 @@ name = "two"
-version = "0.0.0"
+version = "0.0.1"
```

#### test on rust-lnpbp
trying to achieve the same results on `rust-lnpbp`, I checked out commit `064b2f5` (master, tag: v0.2.0-beta.3) and started trying:
- added some changes to all three crates (added a characer to lib.rs's first comment) so cargo ws sees they have changed
- `cargo ws changes -l`
```
info looking for changes since v0.2.0-beta.3
lnpbp          v0.2.0-beta.3 .
lnpbp_derive   v0.2.0-beta.3 derive
lnpbp_services v0.2.0-beta.3
```
- `cargo ws version patch`
```
info looking for changes since v0.2.0-beta.3                                                                                                                                                                                                                                                                                   
info current common version 0.2.0-beta.3                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                               
Changes:                                                                                                                                                                                                                                                                                                                       
 - lnpbp: 0.2.0-beta.3 => 0.2.0                                                                                                                                                                                                                                                                                                
 - lnpbp_derive: 0.2.0-beta.3 => 0.2.0                                                                                                                                                                                                                                                                                         
 - lnpbp_services: 0.2.0-beta.3 => 0.2.0                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                               
✔ Are you sure you want to create these versions? · yes                                                                                                                                                                                                                                                                        
error: There are multiple `lnpbp` packages in your project, and the specification `lnpbp` is ambiguous.                                                                                                                                                                                                                        
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:                                                                                                                                                                                                                                            
  file:///home/dieeasy/git/lnp-bp/rust-lnpbp#lnpbp:0.2.0-beta.3                                                                                                                                                                                                                                                                
  https://github.com/rust-lang/crates.io-index#lnpbp:0.2.0-beta.3                                                                                                                                                                                                                                                              
error: There are multiple `lnpbp_derive` packages in your project, and the specification `lnpbp_derive` is ambiguous.                                                                                                                                                                                                          
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:                                                                                                                                                                                                                                            
  file:///home/dieeasy/git/lnp-bp/rust-lnpbp/derive#lnpbp_derive:0.2.0-beta.3                                                                                                                                                                                                                                                  
  https://github.com/rust-lang/crates.io-index#lnpbp_derive:0.2.0-beta.3                                                                                                                                                                                                                                                       
    Updating crates.io index                                                                                                                                                                                                                                                                                                   
    Updating lnpbp v0.2.0-beta.3 (/home/dieeasy/git/lnp-bp/rust-lnpbp) -> v0.2.0                                                                                                                                                                                                                                               
    Updating lnpbp_derive v0.2.0-beta.3 (/home/dieeasy/git/lnp-bp/rust-lnpbp/derive) -> v0.2.0                                                                                                                                                                                                                                 
    Updating lnpbp_services v0.2.0-beta.3 (/home/dieeasy/git/lnp-bp/rust-lnpbp/services) -> v0.2.0                                                                                                                                                                                                                             
info success ok                                                                                                                                                                                                                                                                                                                
```

resulting diff (missing lnpbp → lnpbp_derive dependency bump):
```diff
diff --git a/Cargo.toml b/Cargo.toml
index efdedc0..19fbe09 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -3 +3 @@ name = "lnpbp"
-version = "0.2.0-beta.3"
+version = "0.2.0"
diff --git a/derive/Cargo.toml b/derive/Cargo.toml
index cc182a0..e49c18d 100644
--- a/derive/Cargo.toml
+++ b/derive/Cargo.toml
@@ -3 +3 @@ name = "lnpbp_derive"
-version = "0.2.0-beta.3"
+version = "0.2.0"
diff --git a/services/Cargo.toml b/services/Cargo.toml
index ae7c336..4f160f0 100644
--- a/services/Cargo.toml
+++ b/services/Cargo.toml
@@ -3 +3 @@ name = "lnpbp_services"
-version = "0.2.0-beta.3"
+version = "0.2.0"
```

as can be seen from the command's output, the `error: There are multiple \`lnpbp\` packages in your project, and the specification `lnpbp` is ambiguous.` may represent an issue

#### further tests
- tried rewriting `lnpbp`'s dependency on `lnpbp_derive` as `lnpbp_derive = { path = "./derive", version = "0.2.0-beta.3" }`, with no results
- tried also rewriting lnpbp_services's dependencies as `lnpbp = { path = "..", version = "0.2.0-beta.3" }` and `lnpbp_derive = { path = "../derive", version = "0.2.0-beta.3" }`, this didn't unlock dependency version updates but removed the "multiple `lnpbp` packages" error
- tried dropping the `lightning-invoice` dependency patch (as it brought `lnpbp-v0.1.0-beta-4` in), `Cargo.lock` cleans up significantly but there's no difference in cargo ws behavior
- tried making the test project as close as possible to lnpbp (main crate + two sub-crates with the same deps and dev-deps between them) and cargo ws correctly updates all crate and dep versions, so no new clues there


#### conclusions
cargo-workspaces looks like the right tool for the job but it's not applicable to the current state of the project

I'll keep digging; in the meantime, if anybody has any info that may help to get cargo ws working as expected (also bump dependency versions), please share

**Comment by dr-orlovsky** [comment]
Thank you @dieeasy for a deep investigation. This looks really strange and as far as I see I can't suggest anything that may help here. Anyway it seems that manual publishing will save much more time than figthing cargo oddness

**Comment by dieeasy** [comment]
as briefly discussed, given that using the `--exact` flag (_Specify inter dependency version numbers exactly with `=`_) is acceptable, I can put together a workflow to version and publish this project automatically, which IMO should still save a lot of time and mistakes with manual publishing

**Comment by dr-orlovsky** [comment]
Solved by https://github.com/LNP-BP/rust-lnpbp/commit/9f5c179e8c81ca50beb6e030c4f4b38a03e05025
